.github/scripts/filter_test_configs.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python3
+
+import sys
+import re
+import json
+import os
+import requests
+from typing import Any, Dict, Set, List
+import yaml
+import warnings
+
+PREFIX = "test-config/"
+SUFFIX = "-only"
+
+# Same as shard names but with the -only suffix to make it clear that
+# only that test config is run
+VALID_TEST_CONFIG_LABELS = {f"{PREFIX}{label}{SUFFIX}" for label in {
+    "backwards_compat",
+    "crossref",
+    "default",
+    "deploy",
+    "distributed",
+    "docs_tests",
+    "dynamo",
+    "force_on_cpu",
+    "functorch",
+    "jit_legacy",
+    "multigpu",
+    "nogpu_AVX512",
+    "nogpu_NO_AVX2",
+    "slow",
+    "xla",
+}}
+
+def parse_args() -> Any:
+    from argparse import ArgumentParser
+    parser = ArgumentParser("Filter all test configurations and keep only requested ones")
+    parser.add_argument("--test-matrix", type=str, required=True, help="the original test matrix")
+    parser.add_argument("--pr-number", type=str, help="the pull request number")
+    parser.add_argument("--tag", type=str, help="the associated tag if it exists")
+    return parser.parse_args()
+
+
+def get_labels(pr_number: int) -> Set[str]:
+    """
+    Dynamical get the latest list of labels from the pull request
+    """
+    # From https://docs.github.com/en/actions/learn-github-actions/environment-variables
+    PYTORCH_REPO = os.environ.get("GITHUB_REPOSITORY", "pytorch/pytorch")
+    PYTORCH_GITHUB_API = f"https://api.github.com/repos/{PYTORCH_REPO}"
+    GITHUB_TOKEN = os.environ["GITHUB_TOKEN"]
+
+    REQUEST_HEADERS = {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": "token " + GITHUB_TOKEN,
+    }
+
+    response = requests.get(
+        f"{PYTORCH_GITHUB_API}/issues/{pr_number}/labels",
+        headers=REQUEST_HEADERS,
+    )
+
+    if response.status_code != requests.codes.ok:
+        warnings.warn(f"Failed to get the labels for #{pr_number} (status code {response.status_code})")
+        return set()
+
+    return {label.get("name") for label in response.json() if label.get("name")}
+
+
+def filter(test_matrix: Dict[str, List[Any]], labels: Set[str]) -> Dict[str, List[Any]]:
+    """
+    Select the list of test config to run from the test matrix. The logic works
+    as follows:
+
+    If the PR has one or more labels as specified in the VALID_TEST_CONFIG_LABELS set, only
+    these test configs will be selected.  This also works with ciflow labels, for example,
+    if a PR has both ciflow/trunk and test-config/functorch-only, only trunk functorch builds
```

Comment:
I'm a bit wary of the `-only` suffix in `test-config/functorch-only`

What if someone wants to run two types of tests in the future?  We'd want to run test that patch either of the two lables, but the label names imply exclusivity E.g. `test-config/dynamo-only` and `test-config/functorch-only`

Reply:
Agree that it sounds a bit counter intuitive having multiple labels with that suffix. We can just call it `test-config/CONFIGNAME`.  The key thing is to communicate clearly what this label does and how to use it safely.  We will keep track of the latter to see how people use it